src/storage/sql_table.cpp

```diff
@@ -19,64 +43,329 @@ SqlTable::SqlTable(BlockStore *const store, const catalog::Schema &schema, const
   TERRIER_ASSERT(attr_sizes.size() == NUM_RESERVED_COLUMNS,
                  "attr_sizes should be initialized with NUM_RESERVED_COLUMNS elements.");
 
+  // First pass through to accumulate the counts of each attr_size
   for (const auto &column : schema.GetColumns()) {
     attr_sizes.push_back(column.GetAttrSize());
   }
 
   auto offsets = storage::StorageUtil::ComputeBaseAttributeOffsets(attr_sizes, NUM_RESERVED_COLUMNS);
 
-  ColumnMap col_oid_to_id;
-  // Build the map from Schema columns to underlying columns
+  ColumnMap col_map;
+  InverseColumnMap inv_col_map;
+
+  // Build the maps between Schema column OIDs and underlying column IDs
   for (const auto &column : schema.GetColumns()) {
     switch (column.GetAttrSize()) {
       case VARLEN_COLUMN:
-        col_oid_to_id[column.GetOid()] = col_id_t(offsets[0]++);
+        inv_col_map[col_id_t(offsets[0])] = column.GetOid();
+        col_map[column.GetOid()] = col_id_t(offsets[0]++);
         break;
       case 8:
-        col_oid_to_id[column.GetOid()] = col_id_t(offsets[1]++);
+        inv_col_map[col_id_t(offsets[1])] = column.GetOid();
+        col_map[column.GetOid()] = col_id_t(offsets[1]++);
         break;
       case 4:
-        col_oid_to_id[column.GetOid()] = col_id_t(offsets[2]++);
+        inv_col_map[col_id_t(offsets[2])] = column.GetOid();
+        col_map[column.GetOid()] = col_id_t(offsets[2]++);
         break;
       case 2:
-        col_oid_to_id[column.GetOid()] = col_id_t(offsets[3]++);
+        inv_col_map[col_id_t(offsets[3])] = column.GetOid();
+        col_map[column.GetOid()] = col_id_t(offsets[3]++);
         break;
       case 1:
-        col_oid_to_id[column.GetOid()] = col_id_t(offsets[4]++);
+        inv_col_map[col_id_t(offsets[4])] = column.GetOid();
+        col_map[column.GetOid()] = col_id_t(offsets[4]++);
         break;
       default:
         throw std::runtime_error("unexpected switch case value");
     }
   }
 
-  auto layout = storage::BlockLayout(attr_sizes);
-  table_ = {new DataTable(block_store_, layout, layout_version_t(0)), layout, col_oid_to_id};
+  // Populate the default value map
+  //
+  // clang's memory analysis gets confused by the iterator here.  This is tied
+  // to the same false positive identified inside of the loop and is caused.
+  // NOLINTNEXTLINE
+  for (const auto &column : schema.GetColumns()) {
+    auto col_oid = column.GetOid();
+    auto *default_value = column.GetDefault();
+    // Only populate the default values of the columns which are new and have a default value
```

Comment:
Is it possible for the user to change the default value of a column? Or else it is treated as a new column (with a different OID)

Reply:
It should be possible for a user to change the default value of a column. But what is happening here is that if a new column is added and a default value is set for that column on creation then it all tuples within the table should receive that default value for that column. But if a default value is set on an already created column then that default value is only applied to tuples inserted from that point on.

Reply:
The storage engine only needs to reason about default values that are applied at the time the column is created as these are retroactive (applied to all existing data).  Default value changes at any other time would only affect new inserts and have to be handled by the catalog, execution engine, and binder because we are not in a position at this level to reason about that granularity of visibility.

Reply:
I think that part deserves better documentation. The user of SqlTable needs to know for default values which part is handled by SqlTable and which part they should take care of.